### PR TITLE
add Framebuffer support to engine

### PR DESCRIPTION
**Pull request by jice-nospam**
Ok so this is the next step towards being able to do render-to-texture / post-processing effects.
It makes it possible to create and bind a framebuffer with an empty texture.

This PR is mainly to discuss how to enable the framebuffer/texture bind using _framebuffer_texture2d_. Right now I added a _with_framebuffer_ bool to _Texture.bind_. I'm not sure that's great.

This still doesn't break unigame's demo and actually allow to render in a 1024x1024 texture as seen here (texture grabbed with RenderDoc) : https://i.imgur.com/8w4DACX.jpg
For some reason, depth test is not enabled. I have to check why. This also stresses the fact that face culling is not currently enabled in the engine.

The next step has more impact on the engine since it should be able to render a screen covering quad using this texture and some fx shader. Not sure yet how to integrate this with the existing game objects / ECS pattern.

### CODE REVIEW

**Comment by edwin0cheng** [comment]
Im thinking about inverting the relationship of the frame buffer object. Instead of holding a texture within a framebuffer, make a new struct named "RenderTexture" and make current Texture struct to become a trait or a container struct which hide the actual impl.

On the other hand, my plan would be added a RenderPass struct to handle a single render pass (pseudo code): 
```
struct RenderPass {
    commands :  Vec<RenderCommand>,
    camera: Camera,
}

struct Camera {
   renderTexture : Option<RenderTexture>
}
```

Im not very good at github skill, is it possible to create a new branch for this feature and working on it together ?

**Comment by edwin0cheng** [comment]
I added a new branch named framebuffer-dev, would you mind to make a pull request there ?

**Comment by edwin0cheng** [comment]
Or do i should push the changes to  jice-nospam/unigame branch ?

**Comment by edwin0cheng** [comment]
Okay , i got how to working on it now...

**Comment by jice-nospam** [comment]
I like the RenderTexture idea. Nice one. Concerning the RenderPass, where do you put the GameObjects ?
I think there are three different cases involving render-to-texture :

- render one scene to a texture, then the same scene from another camera on the screen, using the first texture (for example live screens in a race track)
- render one scene to a texture, then a different scene using the generated texture (for example, live video comm in a game)
- render one scene  to a texture, then render this texture on screen using a post-processing shader (bloom, crt effect,...)

So many we also need another Scene struct in the engine and being able to associate RenderPass with a specific Scene, the postprocessing scene being only a quad covering the screen

**Comment by edwin0cheng** [comment]
For all 3 cases:
* Add render texture to camera, add an method sth like : Camera::Render to render to texture. and use this texture to render to a quad in the main camera. 
* Same as before, except the Camera::Render method take a root node gameobject as argument.
* We would need a Post-Processing Component to handle post processing effect. 

And all these situation idea are shameless stolen how Unity works 😜

**Comment by edwin0cheng** [comment]
And i would pull the current master in first and then continous to change the code.

File: src/engine/render/texture.rs [hunk 1]
```diff
@@ -43,16 +56,18 @@ struct TextureGLState {
 }
 
 impl Texture {
-    pub fn new_empty(width: u32, height: u32) -> Rc<Self> {
+    pub fn new_render_texture(width: u32, height: u32) -> Rc<Self> {
         Rc::new(Texture {
             filtering: TextureFiltering::Linear,
-            img: Resource::new(ImageBuffer::new(width, height)),
             gl_state: RefCell::new(None),
+            kind: TextureKind::RenderTexture(RenderTexture {
```
Comment (jice-nospam) [inline src/engine/render/texture.rs]: I think there's an issue here. The RenderTexture, once generated, has to be used as an Image texture. You need at least a function to turn a renderTexture into a standard texture.

Comment (edwin0cheng) [reply src/engine/render/texture.rs]: You mean render it as a normal texture or get the image buffer itself out to modify directly?

Comment (jice-nospam) [reply src/engine/render/texture.rs]: render it as a normal texture. The steps are :
- create an empty texture (that would be a RenderTexture kind)
- associated it with a framebuffer and bind the framebuffer
- render the scene (to the texture)
- then render something else using the RenderTexture as a standard texture (simply use this texture as a shader parameter). Now, this texture is an Image kind.

**Comment by jice-nospam** [comment]
ok so this can be launch with
`cargo run --example framebuffer --release`

Right now it segfaults, probably because the render texture has no img

**Comment by edwin0cheng** [comment]
Should fix now, just forget to change the coresponding gl.TexImage2D function. (Support 0 len array as null pointer)

**Comment by jice-nospam** [comment]
made a simpler rotating cube example. It starts without framebuffer. Press F1 to turn framebuffer on/off.

**Comment by edwin0cheng** [comment]
I added image widget to display the fb directly in the screen. And changed the demo to use Camera and render_pass api to render the fb. However, because there are no depth buffer attached to it, so the depth test is not working right now.

**Comment by edwin0cheng** [comment]
I added back the cull face method in webgl crate, so right now all things should render properly and webgl version is working too.
You could try it by:
```
cargo web start --example framebuffer
```
By the way, because there are still no game object hierarchy in current engine, all gameobject would be render in the framebuffer, included the framebuffer binded image widget itself. So in the webgl version, there are some warnings in debug console indicated that (in my firefox):
```
Error: WebGL warning: drawElements: Feedback loop detected between tex target 0x0de1, tex unit 2, levels 0-0; and framebuffer attachment 0x8ce0, level 0.
```

It would be fixed after we have a gameobject hierarchy and only render a single gameobject node tree.

**Comment by edwin0cheng** [comment]
So whats the next step ?
As i said before, i would try to make the framebuffer itself holded by RenderTexture, and make the prepare automatically. (It would lazy prepare until Engine::render_pass is called),  So the usage would be something like this:
```
let rt = Rc::new(RenderTexture::new(1024, 1024));
...
let mut cam = game.engine.main_camera.as_ref().unwrap().borrow_mut();
cam.render_texture = Some(rt.clone());
game.engine.render_pass(&cam);
```

Do you have any idea about it?

**Comment by jice-nospam** [comment]
Damn that was quick! I go to sleep and wake up to this! Congratulations :)
I won't be able to code before tomorrow. A few thoughts :

- so we need a Texture trait and two ImageTexture/RenderTexture trait objects
- if prepare is done in render_pass, we have to do it only once. There's no point doing it every frame

**Comment by edwin0cheng** [comment]
* I think i was wrong to use a Texture trait to do that.  I think we could use a Newtype + Deref Idiom to wrap the texture by RenderTexture instead. (Much easier and clean)
* Of course, i did that already in all textures and mesh asset that way. No worry 😉

**Comment by edwin0cheng** [comment]
Yes, the post-processing effect should be using another mechanism instead of ui widget. For other game objects, an material component could be assigned to use custom shader.

**Comment by edwin0cheng** [comment]
I would settle here for this usage (rt.as_texture()) now. And if there are any other usage, then just refactor it to use a Texture trait. and i would merge this to the master.

**Pull request merged**